processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -267,55 +291,160 @@ static double estimatePredicateSelectivity(
   }
 
   static double estimatePredicateSelectivity(
+      BitmapIndex bitmapIndex,
       ColumnSelector columnSelector,
       String dimension,
-      Iterable<ImmutableBitmap> bitmaps,
+      Iterable<Integer> bitmapIndexes,
       long totalNumRows
   )
   {
     final ColumnCapabilities columnCapabilities = columnSelector.getColumn(dimension).getCapabilities();
-    return estimatePredicateSelectivity(
-        bitmaps,
+    return estimateSelectivityOfBitmapList(
+        bitmapIndex,
+        bitmapIndexes,
         totalNumRows,
+        // assume multi-value column if columnCapabilities is null
         columnCapabilities == null || columnCapabilities.hasMultipleValues()
     );
   }
 
-  static double estimatePredicateSelectivity(
+  private static double estimateSelectivityOfBitmapList(
+      BitmapIndex bitmapIndex,
+      Iterable<Integer> bitmapIndexeIterable,
+      long totalNumRows,
+      boolean isMultiValueDimension
+  )
+  {
+    final List<Integer> bitmapIndexes = ImmutableList.copyOf(bitmapIndexeIterable);
+    long numMatchedRows = 0;
+    for (Integer index : bitmapIndexes) {
+      final ImmutableBitmap bitmap = bitmapIndex.getBitmap(index);
+      numMatchedRows += bitmap.size();
```

Comment:
Right, size() is not cheap. So, you mean frequent call of size() is expensive? I'll caching it.